### ISSUE TITLE
Equal mixture proportions

### ISSUE THREAD

**Issue by linatakemaru** [mahmoodlab/PANTHER#23]
Hi, thanks for your awesome work! I tried to implement PANTHER on around 3000 WSIs. I successfully generated a prototype and embedding for c=10 and c=16, and ran the demo code for visualization. However, when I calculate the "mus", they're almost all equal for every WSI (see below). Any thoughts on why this is happening? 

I also noticed that the imbalance value during training is consistently equal to k (16), which I believe is incorrect.

Thank you! 

<img width="511" alt="Screenshot 2024-10-30 at 1 30 14 PM" src="https://github.com/user-attachments/assets/eb342500-33ee-4f84-9cc6-f598fc2da907">

**Comment by andrewsong90** [comment]
Hi @linatakemaru 

I am not sure why this could be happening - Have you checked each prototype to see if they are sufficiently different?
One more solution could be to normalize all embeddings so that L2 norm is 1 and then trying the procedure again to see if it solves the issue.

Which patch encoder are you using?

**Comment by linatakemaru** [comment]
Hi, thanks for the response. I'm using UNI–I tried redownloading the model from huggingface and generated the embeddings from scratch again, but ran into the same issue. Here's a screenshot of the output for more context. 

I believe this has something to do with the faiss implementation of kmeans. When I switch the `mode` parameter from `faiss` to `kmeans` to use the sklearn CPU version, the visualizations and example patches from each cluster make sense morphologically. However, since sklearn doesn't output the same metrics per iteration, I'm not sure if those metric values are expected. For now, I've resorted to using sklearn until it gets sorted out.

<img width="549" alt="image" src="https://github.com/user-attachments/assets/5087bec6-5363-4bc9-9880-1bcd12f7c8ec">
